crates/wasmtime/src/runtime/gc/enabled/exnref.rs

```diff
@@ -0,0 +1,784 @@
+        assert!(self.comes_from_same_store(&store));
+        let type_index = self.type_index(store)?;
+        let layout = store
+            .engine()
+            .signatures()
+            .layout(type_index)
+            .expect("exn types should have GC layouts");
+        match layout {
+            GcLayout::Struct(_) => unreachable!(),
+            GcLayout::Array(_) => unreachable!(),
+            GcLayout::Exception(e) => Ok(e),
+        }
+    }
+
+    fn field_ty(&self, store: &StoreOpaque, field: usize) -> Result<FieldType> {
+        let ty = self._ty(store)?;
+        match ty.field(field) {
+            Some(f) => Ok(f),
+            None => {
+                let len = ty.fields().len();
+                bail!("cannot access field {field}: exn only has {len} fields")
+            }
+        }
+    }
+
+    /// Get this exception object's `index`th field.
+    ///
+    /// # Errors
+    ///
+    /// Returns an `Err(_)` if the index is out of bounds or this reference has
+    /// been unrooted.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this reference is associated with a different store.
+    pub fn field(&self, mut store: impl AsContextMut, index: usize) -> Result<Val> {
+        let mut store = AutoAssertNoGc::new(store.as_context_mut().0);
+        self._field(&mut store, index)
+    }
+
+    pub(crate) fn _field(&self, store: &mut AutoAssertNoGc<'_>, index: usize) -> Result<Val> {
+        assert!(self.comes_from_same_store(store));
+        let exnref = self.exnref(store)?.unchecked_copy();
+        let field_ty = self.field_ty(store, index)?;
+        let layout = self.layout(store)?;
+        Ok(exnref.read_field(store, &layout, field_ty.element_type(), index))
+    }
+
+    /// Get this exception object's associated tag.
+    ///
+    /// # Errors
+    ///
+    /// Returns an `Err(_)` if this reference has been unrooted.
+    ///
+    /// # Panics
+    ///
+    /// Panics if this reference is associated with a different store.
+    pub fn tag(&self, mut store: impl AsContextMut) -> Result<Tag> {
+        let mut store = AutoAssertNoGc::new(store.as_context_mut().0);
+        assert!(self.comes_from_same_store(&mut store));
+        let exnref = self.exnref(&mut store)?.unchecked_copy();
+        let layout = self.layout(&mut store)?;
+        let (instance, index) = exnref.tag(&mut store, &layout)?;
+        Ok(Tag::from_raw_indices(&mut *store, instance, index))
+    }
+}
+
+unsafe impl WasmTy for Rooted<ExnRef> {
+    #[inline]
+    fn valtype() -> ValType {
+        ValType::Ref(RefType::new(false, HeapType::Exn))
+    }
+
+    #[inline]
+    fn compatible_with_store(&self, store: &StoreOpaque) -> bool {
+        self.comes_from_same_store(store)
+    }
+
+    #[inline]
+    fn dynamic_concrete_type_check(
+        &self,
+        store: &StoreOpaque,
+        _nullable: bool,
+        ty: &HeapType,
+    ) -> Result<()> {
+        self.ensure_matches_ty(store, ty)
```

Comment:
Technically this should alwayas be `Ok(())`, right? Given that it's impossible for a wasm module to request a typed exnref?